File: DSAlgorithms/Leetcode/c++/arrays_hash/2001-interchangeable_rectangles.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

    long long interchangeableRectangles(std::vector<std::vector<int>>& rectangles) {
        std::unordered_map<long double, int> hash;    
        long double ratio = 0;
        long long result = 0;
        for(int i = 0; i < rectangles.size(); i++)
        {
            ratio = (long double) (rectangles[i][0])/ (long double) (rectangles[i][1]);
            if(hash.count(ratio) > 0)
                hash[ratio]++;
            else 
                hash[ratio] = 1;
        }
        for(auto item: hash)
        {
            result += (long long) (item.second) * (long long) (item.second - 1) / 2;
        }
        return result;
    }

File: dsa_algo/leetcode/ts/src/arrays_hash/2001-interchangeable_rectangles.ts
const interchangeableRectangles = (rectangles: number[][]): number =>{
    const hash = new Map<number, number>();
    let ratio: number = 0;
    let result: number = 0;
    for(let i = 0; i < rectangles.length; i++)
    {
        ratio = rectangles[i][0]/rectangles[i][1];
        if(hash.has(ratio))
        {
            const current_val = hash.get(ratio);
            hash.set(ratio, current_val + 1);
        }
        else
            hash.set(ratio, 1);
    }
    hash.forEach((value, key) => {
        if(value > 1)
            result += (value) * (value - 1) / 2;
    })
    return result;
};
